### PR TITLE
pallet-xcm: add new flexible `transfer_assets()` call/extrinsic

### PR DESCRIPTION
**Pull request by acatangiu**
# Motivation (+testing)

### Enable easy `ForeignAssets` transfers using `pallet-xcm` 

We had just previously added capabilities to teleport fees during reserve-based transfers, but what about reserve-transferring fees when needing to teleport some non-fee asset?

This PR aligns everything under either explicit reserve-transfer, explicit teleport, or this new flexible `transfer_assets()` which can mix and match as needed with fewer artificial constraints imposed to the user.

This will enable, for example, a (non-system) parachain to teleport their `ForeignAssets` assets to AssetHub while using DOT to pay fees.
(the assets are teleported - as foreign assets should from their owner chain - while DOT used for fees can only be reserve-based transferred between said parachain and AssetHub).

Added `xcm-emulator` tests for this scenario ^.

# Description

Reverts `(limited_)reserve_transfer_assets` to only allow reserve-based transfers for all `assets` including fees.

Similarly `(limited_)teleport_assets` only allows teleports for all `assets` including fees.
    
For complex combinations of asset transfers where assets and fees may have different reserves or different reserve/teleport trust configurations, users can use the newly added `transfer_assets()` extrinsic which is more flexible in allowing more complex scenarios.

`assets` (excluding `fees`) must have same reserve location or otherwise be teleportable to `dest`.
No limitations imposed on `fees`.

- for local reserve: transfer assets to sovereign account of destination chain and forward a notification XCM to `dest` to mint and deposit reserve-based assets to `beneficiary`.
- for destination reserve: burn local assets and forward a notification to `dest` chain to withdraw the reserve assets from this chain's sovereign account and deposit them to `beneficiary`.
- for remote reserve: burn local assets, forward XCM to reserve chain to move reserves from this chain's SA to `dest` chain's SA, and forward another XCM to `dest` to mint and deposit reserve-based assets to `beneficiary`.
- for teleports: burn local assets and forward XCM to `dest` chain to mint/teleport assets and deposit them to `beneficiary`.

## Review notes

Only around 500 lines are prod code (see `pallet_xcm/src/lib.rs`), the rest of the PR is new tests and improving existing tests.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4382996

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4382994

**Comment by acatangiu** [comment]
@KiChjang @franciscoaguirre PTAL and let me know if there's arguments against it - if we are aligned, I will then continue with adding the missing pieces to the PR.

**Comment by franciscoaguirre** [comment]
I like not requiring the user to think about teleports or reserve backed transfers, while still keeping the other extrinsics for more specific use-cases and for devs that have to think about this stuff

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4519118

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4519117

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4551592

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4551593

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4551594

**Comment by acatangiu** [comment]
@KiChjang @franciscoaguirre this is good to go, please review

File: cumulus/parachains/integration-tests/emulated/common/src/macros.rs [hunk 1]
```diff
@@ -120,3 +120,107 @@ macro_rules! test_parachain_is_trusted_teleporter {
...
+					asset_id_on_penpal,
+					ASSET_MIN_BALANCE,
+					is_sufficient,
+					asset_owner,
+					None,
+					prefund_amount,
+				);
+
+				let require_weight_at_most = Weight::from_parts(1_100_000_000_000, 30_000);
+				let origin_kind = OriginKind::Xcm;
```
Comment (franciscoaguirre) [inline cumulus/parachains/integration-tests/emulated/common/src/macros.rs:171]: Shouldn't this be the `SovereignAccount` origin kind? It's the equivalent of the `Signed` FRAME origin

File: cumulus/parachains/integration-tests/emulated/common/src/macros.rs [hunk 2]
```diff
@@ -120,3 +120,107 @@ macro_rules! test_parachain_is_trusted_teleporter {
...
+				let sov_penpal_on_ah_as_location = MultiLocation {
+					parents: 0,
+					interior: X1(
+						AccountId32Junction { network: None, id: sov_penpal_on_ah.clone().into() }
+					),
+				};
```
Comment (franciscoaguirre) [inline cumulus/parachains/integration-tests/emulated/common/src/macros.rs:177]: You don't need to create this location, just use the location for the chain and the sovereign account will be calculated correctly to deposit it there.

File: polkadot/xcm/pallet-xcm/src/lib.rs [hunk 3]
```diff
@@ -1225,12 +1252,162 @@ pub mod pallet {
...
 			Ok(())
 		}
+
+		/// Transfer some assets from the local chain to the destination chain through their local,
+		/// destination or remote reserve, or through teleports.
+		///
+		/// Fee payment on the destination side is made from the asset in the `assets` vector of
+		/// index `fee_asset_item` (hence referred to as `fees`), up to enough to pay for
+		/// `weight_limit` of weight. If more weight is needed than `weight_limit`, then the
+		/// operation will fail and the assets send may be at risk.
```
Comment (franciscoaguirre) [inline polkadot/xcm/pallet-xcm/src/lib.rs:1262]:
```suggestion
		/// operation will fail and the assets sent may be at risk.
```

**Comment by acatangiu** [comment]
bot bench cumulus-assets --runtime=asset-hub-rococo --pallet=pallet_xcm

File: cumulus/parachains/integration-tests/emulated/common/src/macros.rs (hunk 1 above)
Comment (acatangiu) [reply cumulus/parachains/integration-tests/emulated/common/src/macros.rs:171]: Has to be `Xcm` - imposed by `ForeignCreators` to only allow creating foreign assets through XCM: https://github.com/paritytech/polkadot-sdk/blob/bfb241d7f3e29152b9fe0ab3a0364ee0c89ec65c/cumulus/parachains/runtimes/assets/common/src/foreign_creators.rs#L45C15-L45C15

**Comment by acatangiu** [comment]
bot clean

**Review by svyatonik — APPROVED** [review]
Mostly have been looking at `pallet-xcm` changes - looks good!

File: polkadot/xcm/pallet-xcm/src/lib.rs [hunk 4]
```diff
@@ -1341,64 +1532,26 @@ impl<T: Config> Pallet<T> {
...
-					return Err(Error::<T>::InvalidAssetUnsupportedReserve.into()),
-			});
-		};
+		// Find transfer types for fee and non-fee assets.
+		let (fees_transfer_type, assets_transfer_type) =
+			Self::find_fee_and_assets_transfer_types(&assets, fee_asset_item, &dest)?;
+		// Ensure assets (and fees according to check below) are not teleportable to `dest`.
+		ensure!(assets_transfer_type != TransferType::Teleport, Error::<T>::Filtered);
+		// Ensure all assets (including fees) have same reserve location.
+		ensure!(assets_transfer_type == fees_transfer_type, Error::<T>::TooManyReserves);
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply polkadot/xcm/pallet-xcm/src/lib.rs:1546]: there is a note for it already in prdoc ; added label too

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v1-5-0/5358/1

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-kusama-bridge/2971/8

**Comment by brenzi** [comment]
This works for our parachain :+1: 
However: can it be that teleporting our native token to asset hub only works if the order of assets in the `transferAsset` call is (TEER, ROC, feeAsset = 1)?
I first tried (ROC, TEER, feeAsset=0) but that didn't work

**Comment by joepetrowski** [comment]
I will leave it to @acatangiu to answer, but that does seem strange to me.

**Comment by acatangiu** [comment]
> This works for our parachain 👍 However: can it be that teleporting our native token to asset hub only works if the order of assets in the `transferAsset` call is (TEER, ROC, feeAsset = 1)? I first tried (ROC, TEER, feeAsset=0) but that didn't work

Like the other extrinsics, `transferAsset()` takes the assets to transfer through a [Box<VersionedMultiAssets>](https://github.com/paritytech/polkadot-sdk/blob/61801ce324f18eacb9c29e2f674f0d262a00bd4d/polkadot/xcm/pallet-xcm/src/lib.rs#L1314C12-L1314C37) parameter.

`MultiAssets` object [internally sorts](https://github.com/paritytech/polkadot-sdk/blob/5fc7622cb312f2d32ec8365012ee0a49622db8c8/polkadot/xcm/src/v3/multiasset.rs#L691) the list of assets (I'm not sure where this requirement to be sorted comes from, maybe required for encoding/decoding?).

So my guess is you're calling doing something like:
```rust
let assets = vec![TEER, ROC].into();    // <-- this works
let fee_idx = 1;                        // ok

let assets = vec![ROC, TEER].into();    // <-- still works, but `fee_idx` should still be `1` because of sorting :(
let fee_idx = 0; // <-- reading this looks like it's pointing to ROC, but because of the sorting done by `.into()`, TEER is actually idx 0 and ROC is idx 1
```
(local assets will always go first because of parents=0 vs ROC which has parents=1)

and I know... this is annoying and bad API design - we could change extrinsic to take actual fee asset instead of index, but would be a breaking change..

Also here is an [example helper function](https://github.com/paritytech/polkadot-sdk/blob/5fc7622cb312f2d32ec8365012ee0a49622db8c8/polkadot/xcm/pallet-xcm/src/tests/assets_transfer.rs#L288) I used in tests to get the right index for transfers.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/xcm-transfer-reserve-non-reserve-assets-in-a-single-xcm/2808/8

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/xcm-user-and-developer-experience-improvements/4511/21

**Comment by gitkit13** [comment]
> # Motivation (+testing)
> ### Enable easy `ForeignAssets` transfers using `pallet-xcm`
> We had just previously added capabilities to teleport fees during reserve-based transfers, but what about reserve-transferring fees when needing to teleport some non-fee asset?
> 
> This PR aligns everything under either explicit reserve-transfer, explicit teleport, or this new flexible `transfer_assets()` which can mix and match as needed with fewer artificial constraints imposed to the user.
> 
> This will enable, for example, a (non-system) parachain to teleport their `ForeignAssets` assets to AssetHub while using DOT to pay fees. (the assets are teleported - as foreign assets should from their owner chain - while DOT used for fees can only be reserve-based transferred between said parachain and AssetHub).
> 
> Added `xcm-emulator` tests for this scenario ^.
> 
> # Description
> Reverts `(limited_)reserve_transfer_assets` to only allow reserve-based transfers for all `assets` including fees.
> 
> Similarly `(limited_)teleport_assets` only allows teleports for all `assets` including fees.
> 
> For complex combinations of asset transfers where assets and fees may have different reserves or different reserve/teleport trust configurations, users can use the newly added `transfer_assets()` extrinsic which is more flexible in allowing more complex scenarios.
> 
> `assets` (excluding `fees`) must have same reserve location or otherwise be teleportable to `dest`. No limitations imposed on `fees`.
> 
> * for local reserve: transfer assets to sovereign account of destination chain and forward a notification XCM to `dest` to mint and deposit reserve-based assets to `beneficiary`.
> * for destination reserve: burn local assets and forward a notification to `dest` chain to withdraw the reserve assets from this chain's sovereign account and deposit them to `beneficiary`.
> * for remote reserve: burn local assets, forward XCM to reserve chain to move reserves from this chain's SA to `dest` chain's SA, and forward another XCM to `dest` to mint and deposit reserve-based assets to `beneficiary`.
> * for teleports: burn local assets and forward XCM to `dest` chain to mint/teleport assets and deposit them to `beneficiary`.
> 
> ## Review notes
> Only around 500 lines are prod code (see `pallet_xcm/src/lib.rs`), the rest of the PR is new tests and improving existing tests.

gh pr checkout 2062

**Comment by gitkit13** [comment]
> > This works for our parachain 👍 However: can it be that teleporting our native token to asset hub only works if the order of assets in the `transferAsset` call is (TEER, ROC, feeAsset = 1)? I first tried (ROC, TEER, feeAsset=0) but that didn't work
> 
> Like the other extrinsics, `transferAsset()` takes the assets to transfer through a [Box<VersionedMultiAssets>](https://github.com/paritytech/polkadot-sdk/blob/61801ce324f18eacb9c29e2f674f0d262a00bd4d/polkadot/xcm/pallet-xcm/src/lib.rs#L1314C12-L1314C37) parameter.
> 
> `MultiAssets` object [internally sorts](https://github.com/paritytech/polkadot-sdk/blob/5fc7622cb312f2d32ec8365012ee0a49622db8c8/polkadot/xcm/src/v3/multiasset.rs#L691) the list of assets (I'm not sure where this requirement to be sorted comes from, maybe required for encoding/decoding?).
> 
> So my guess is you're doing something like:
> 
> ```rust
> let assets: MultiAssets = vec![TEER, ROC].into();    // <-- this works
> let fee_idx = 1;                                     // ok
> 
> let assets: MultiAssets = vec![ROC, TEER].into();    // <-- still works, but `fee_idx` should still be `1` because of sorting :(
> let fee_idx = 0; // <-- reading this looks like it's pointing to ROC, but because of the sorting done by `.into()`, TEER is actually idx 0 and ROC is idx 1
> ```
> 
> (local assets will always go first because of parents=0 vs ROC which has parents=1)
> 
> and I know... this is annoying and bad API design - we could change extrinsic to take actual fee asset instead of index, but would be a breaking change..
> 
> Also here is an [example helper function](https://github.com/paritytech/polkadot-sdk/blob/5fc7622cb312f2d32ec8365012ee0a49622db8c8/polkadot/xcm/pallet-xcm/src/tests/assets_transfer.rs#L288) I used in tests to get the right index for transfers.

**Pull request merged**